mongodbatlas/fw_data_source_mongodbatlas_atlas_users.go

```diff
@@ -0,0 +1,256 @@
+							Computed: true,
+						},
+						"mobile_number": schema.StringAttribute{
+							Computed: true,
+						},
+						"team_ids": schema.ListAttribute{
+							ElementType: types.StringType,
+							Computed:    true,
+						},
+						"links": schema.ListNestedAttribute{
+							Computed: true,
+							NestedObject: schema.NestedAttributeObject{
+								Attributes: map[string]schema.Attribute{
+									"href": schema.StringAttribute{
+										Computed: true,
+									},
+									"rel": schema.StringAttribute{
+										Computed: true,
+									},
+								},
+							},
+						},
+						"roles": schema.ListNestedAttribute{
+							Computed: true,
+							NestedObject: schema.NestedAttributeObject{
+								Attributes: map[string]schema.Attribute{
+									"group_id": schema.StringAttribute{
+										Computed: true,
+									},
+									"org_id": schema.StringAttribute{
+										Computed: true,
+									},
+									"role_name": schema.StringAttribute{
+										Computed: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (d *AtlasUsersDS) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	connV2 := d.client.AtlasV2
+
+	var atlasUsersConfig tfAtlasUsersDSModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &atlasUsersConfig)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if atlasUsersConfig.OrgID.IsNull() && atlasUsersConfig.ProjectID.IsNull() {
+		resp.Diagnostics.AddError(errorMissingAttributesSummary, errorMissingAttributesDetail)
+		return
+	}
+
+	var (
+		users      []admin.CloudAppUser
+		totalCount int
+	)
+
+	switch {
+	case !atlasUsersConfig.ProjectID.IsNull():
+		projectID := atlasUsersConfig.ProjectID.ValueString()
+		apiResp, _, err := connV2.ProjectsApi.ListProjectUsersWithParams(ctx, &admin.ListProjectUsersApiParams{
+			GroupId:      projectID,
+			PageNum:      int64PtrToIntPtr(atlasUsersConfig.PageNum.ValueInt64Pointer()),
+			ItemsPerPage: int64PtrToIntPtr(atlasUsersConfig.ItemsPerPage.ValueInt64Pointer()),
+		}).Execute()
+		if err != nil {
+			resp.Diagnostics.AddError("error when getting users from Atlas", fmt.Sprintf(errorUsersRead, "project", projectID, err.Error()))
+			return
+		}
+		users = apiResp.Results
+		totalCount = *apiResp.TotalCount
+	case !atlasUsersConfig.TeamID.IsNull() && !atlasUsersConfig.OrgID.IsNull():
+		teamID := atlasUsersConfig.TeamID.ValueString()
+		apiResp, _, err := connV2.TeamsApi.ListTeamUsersWithParams(ctx, &admin.ListTeamUsersApiParams{
+			OrgId:        atlasUsersConfig.OrgID.ValueString(),
+			TeamId:       teamID,
+			PageNum:      int64PtrToIntPtr(atlasUsersConfig.PageNum.ValueInt64Pointer()),
+			ItemsPerPage: int64PtrToIntPtr(atlasUsersConfig.ItemsPerPage.ValueInt64Pointer()),
+		}).Execute()
+		if err != nil {
+			resp.Diagnostics.AddError("error when getting users from Atlas", fmt.Sprintf(errorUsersRead, "team", teamID, err.Error()))
+			return
+		}
```

Comment:
this error is replicated in all the cases. Can we move it outside the switch?

Reply:
Gave this a try. Each error does have a specific message using the particular id that was used, as well as the non-nil `err`. Becomes quite complex to extract defeating the purpose of simplifying.